### DISCUSSION TITLE
Deployment Help

### DISCUSSION THREAD

**Discussion by tonydean37** [Azure/ipam#90] [Q&A] [answered]
Hi @DCMattyG, I got permission to deploy the solution, so have done so, but we have a bit of an issue regarding accessing the key vault. Our policy currently dictates that we can only access a keyvault over private IP, so would normally create a private link to enable this access.

Looking at the Service App, this could be done using vnet integration:
https://learn.microsoft.com/en-gb/azure/app-service/overview-vnet-integration

But the option is greyed out, stating:
Virtual network integration is only supported on Basic and higher plans. Virtual network integration is not supported with multi-container (Docker Compose) apps.

I believe the issue here is that this is a multi-container (Docker Compose) app. I wondered if you knew of another way around this issue?

Cheers in advance!

**Comment by DCMattyG** [comment]
Hey there @tonydean37! You are correct, at this moment multi-tenant App Services do not support Virtual Network integrations 😢

For other customers whom have similar requirements, I've seen them go one of two ways...

1) **AKS:** We have another customer whom posted their solution for deploying Azure IPAM in AKS, and they were kind enough to share their deployment configurations and settings (See #32).
2) **Separate App Services behind App Gateway:** Many customers have chosen to go down this path as it allows the most flexibility while leveraging all Azure-native components. The only "gotcha" per-se is that you need to bring your own SSL cert, whereas with the multi-container App Service has an SSL cert integrated.

Here is a diagram of the App Gateway solution (example):

![image](https://user-images.githubusercontent.com/12955732/222474007-fc140779-1ed8-4991-a627-6be68005382d.png)

So...after digesting that a bit, please let me know your thoughts on the above, and what direction you think would be best for you. It's ok if it's not one that I listed above, we're work together to figure it out!

**Reply by tonydean37** [reply]
Hi @DCMattyG, As always I really appreciate the quick response. I've had sign off to create an exemption for the Cosmos DB and Key Vault traffic flows while in PoC, which will give us some time to evaluate the best option moving into production over the coming weeks. I'll keep you posted with our progress. 

Have a great weekend!

**Comment by tonydean37** [comment, marked as answer]
I've come across an issue with our deployment (we had to change some of the scripts due to policy and trying to keep the scope limited to a certain resource management group rather than our whole tenant). 

I am getting an error message when trying to access the website and see the below error messages in the logs and I wondered if you could help?
![image](https://user-images.githubusercontent.com/116811475/222690848-b2168cf7-72b4-42e9-a290-48287005b934.png)

2023-03-03T09:56:16.316856035Z Content: {"error":"unauthorized_client","error_description":"AADSTS700016: Application with identifier 'System.Collections.Hashtable' was not found in the directory 'Hastings Insurance Services Ltd.'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant.\r\nTrace ID: 0d1bf8ba-5698-437e-aae7-3823593b0700\r\nCorrelation ID: a662b8dc-2cd4-4de7-b2e6-474e791b651c\r\nTimestamp: 2023-03-03 09:55:16Z","error_codes":[700016],"timestamp":"2023-03-03 09:55:16Z","trace_id":"0d1bf8ba-5698-437e-aae7-3823593b0700","correlation_id":"a662b8dc-2cd4-4de7-b2e6-474e791b651c","error_uri":"https://login.microsoftonline.com/error?code=700016"}

I didn't know if it was in relation to a change that was made to the deploy.ps1 file here:
![image](https://user-images.githubusercontent.com/116811475/222716570-defe99b0-a036-4a4e-9f87-ff9e01013989.png)

We amended the $TenantId to the name of the management group that we wanted to limit the scope of the work too.

**Reply by tonydean37** [reply]
Morning @DCMattyG, I hope you had a good weekend. I realised I didn't tag you in the message above, so doing so now. :)

The other thing I thought I should mention, when we deployed the resources I had owner over the sub-management group and I was provided Application Admin rather than Global Admin, due to policy. The deployment seemed to run without any errors though.

Cheers.

**Reply by DCMattyG** [reply]
Hi @tonydean37 , it would be helpful to know how the script was modified to provide a better answer. 

Global Administrator is required because Admin Consent needs to be set for the Application Registrations. Global Admins can set this to allow the application permissions at the Tenant level, and the script attempts to do this on your behalf.

At face value, the error seems to imply Admin Consent is not set, but there may be more going on depending on the changes that were made.

I'm more that happy to setup a 1:1 with you to help you troubleshoot in detail, and come up with a deployment that works for the constraints of your organization. Don't hesitate to reach out to me via email at Matthew.Garrett@microsoft.com anytime.

**Reply by tonydean37** [reply]
Morning @DCMattyG, I thought I should let you know that we have resolved the issues we were having with the deployment. :) 

FYI - we used owner on our test management group, which is a sub group on our main tenant. We also used Application Admin instead of Global Admin to adhere to our least privilege policy and amended the $scope to be our test environment instead of $TenantId, which seems to have worked as we can see our VNets, Subnets and Resources.

I'm sure I'll have some questions over the coming days/weeks, so may give you an email to setup a meeting. Thanks again for all the help so far. :)

**Reply by DCMattyG** [reply]
This is wonderful news @tonydean37. Pease keep me posted on how things work, and if you run into any issues. I'm also interested in documenting the process you followed so others whom would like to use lower-privileged accounts can follow in your footsteps.

Thanks so much!

**Reply by tonydean37** [reply]
Morning @DCMattyG, Sure thing. The one part of the script we amended was this one:
   $scope = "/providers/Microsoft.Management/managementGroups/[instead of $TenantId we used the name of the management group we wanted to restrict the permissions over]"

Then just ran it with the Azure Resource role: Owner for that management group (we didn't try Contributor) and the Azure AD role: Application Administrator. Happy to go through it on a call at some point in the future if you'd like though. :)

I've come across something while trying to get the subnet script working and wondered if you could assist? I've been able to successfully run the VNet reservation script, and see the VNET reservations under the Block, but am getting an error back when trying to adapt the script for the next available subnet. Here is the script I'm using:

$engineClientId = '[engineId]'
$appName = '[ipam-xxxx]'
$vnet = '[Virtual network ID - taken from virtual network page in Azure]'
$size = 28

$accessToken = ConvertTo-SecureString ($(az account get-access-token --resource=api://$engineClientId --query accessToken --output tsv)) -AsPlainText

$requestUrl = "https://$appName.azurewebsites.net/api/tools/nextAvailableSubnet"

$bodysubnet = @{
  'vnet_id' = $vnet
  'size' = $size
  'reverse_search' = "false"
  'smallest_cidr' = "false"
} | ConvertTo-Json

$bodysubnet

$headers = @{
  'Accept' = 'application/json'
  'Content-Type' = 'application/json'
}

$responsesubnet = Invoke-RestMethod `
 -Method 'Post' `
 -Uri $requestUrl `
 -Authentication 'Bearer' `
 -Token $accessToken `
 -Headers $headers `
 -Body $bodysubnet

 $responsesubnet


----------------
I get the following from $bodysubnet:

{
  "size": 28,
  "reverse_search": "false",
  "vnet_id": "[correct VNET ID]",
  "smallest_cidr": "false"
}

---------------
I get the following error message:

Invoke-RestMethod: C:\repos\demo-repo\ipam-Subnet-prefix.ps1:25:19
Line |
  25 |  $responsesubnet = Invoke-RestMethod `
     |                    ~~~~~~~~~~~~~~~~~~~
     | {"error":"Invalid Virtual Network ID."}

I've tried changing the VNET ID to the name of the VNET, but I get the same result. I wondered if you could advise where I've missed something?

Cheers!!

**Reply by DCMattyG** [reply]
Hey @tonydean37 , the ID is the Virtual Network *Resource* ID, which can be found on the Properties page of the Virtual Network. I'll update the API documentation to make that more clear. 

Hope that helps!

**Reply by tonydean37** [reply]
Ideal - thank you :)

**Reply by esdcharles** [reply]
> Morning @DCMattyG, I thought I should let you know that we have resolved the issues we were having with the deployment. :)
> 
> FYI - we used owner on our test management group, which is a sub group on our main tenant. We also used Application Admin instead of Global Admin to adhere to our least privilege policy and amended the $scope to be our test environment instead of $TenantId, which seems to have worked as we can see our VNets, Subnets and Resources.
> 
> I'm sure I'll have some questions over the coming days/weeks, so may give you an email to setup a meeting. Thanks again for all the help so far. :)

I am in a very similar situation in which we are being asked to adhere to our least privilege policy as well. Thanks for this post @tonydean37 as I'm sure many others will benefit from your approach to this. Stay tuned with the hope my outcome is the same as yours.

**Comment by jeevanions** [comment]
@tonydean37 Have you tried using Application OwnedBy permission instead of Application Administrator? Would this work?

**Reply by tonydean37** [reply]
Hi @jeevanions - it's not one we tried to be honest. It's not a permission I have used before, so may be worth doing some testing with it to see what access it gives etc...
